### PR TITLE
linux: teach uv_get_constrained_memory() cgroupsv2

### PR DESCRIPTION
**Pull request by bnoordhuis**
Fixes: https://github.com/libuv/libuv/issues/2315

cc @maleadt - I figured it'd be easier to type it out than having you jump through more hoops.

### CODE REVIEW

File: src/unix/linux-core.c
Comment (bnoordhuis) [inline src/unix/linux-core.c:831]: This is the only part I'm not really sure about. I think there can be more lines but I suspect the first line is always the root hierarchy. Please correct me if I'm wrong.

Comment (maleadt) [reply src/unix/linux-core.c:831]: This should be fine for cgroupv2 detection. According to kernel docs:

> `/proc/$PID/cgroup` lists a process's cgroup membership.  If legacy
> cgroup is in use in the system, this file may contain multiple lines,
> one for each hierarchy.  The entry for cgroup v2 is always in the
> format `0::$PATH`:
>
> ```
>  # cat /proc/842/cgroup
>  ...
>  0::/test-cgroup/test-cgroup-nested
> ```

**Review by maleadt — APPROVED** [review]
The simplification here is that you assume cgroupv2 to be mounted at `/sys/fs/cgroup`, and avoid parsing `mountinfo`. That's not technically correct, but then again I haven't seen `cgroupv2` be mounted anywhere else, so maybe it's fine to go with the simpler solution until somebody complains. The only exception is the hybrid cgroupv1+cgroupv2 set-up, which mounted the v2 hierarchy at `/sys/fs/cgroup/unified`, but that was a transitional solution, and the fallback still works.

**Comment by maleadt** [comment]
So do we go ahead with this PR, @bnoordhuis?

**Comment by bnoordhuis** [comment]
I'm waiting for a review from @vtjnash or another maintainer.

File: src/unix/linux-core.c
Comment (vtjnash) [inline src/unix/linux-core.c:821]:
```suggestion
  char filename[1200]; /* slightly larger than sizeof(buf) */
  char buf[1024];
```

**Review by vtjnash — APPROVED** [review]
This SGTM

**Comment by maleadt** [comment]
Great, thanks @bnoordhuis!

In addition to `memory.max`/`memory.high`, we probably also need a way to expose `memory.current` (because `uv_get_free_memory` currently does not reflect any cgroup constraint). What about adding `uv_get_available_memory` that queries `memory.current` or falls back to `uv_get_free_memory` (if unconstrained)? It seems like that would pair well: `uv_get_memory_constrained`/`uv_get_memory_available` cfr. `uv_get_memory_total`/`uv_get_memory_free`.

File: src/unix/linux-core.c
Comment (maleadt) [inline src/unix/linux-core.c:804]: @bnoordhuis I just noticed that this returns -1 when no limit is set. The documentation currently states that 0 is returned for both no and unknown limits. Do you want to keep it like this (being able to discern between both cases) and update the documentation, or return 0 from here as well?

Comment (bnoordhuis) [reply src/unix/linux-core.c:804]: The documentation is ambiguously worded.

"If there is no such constraint, or the constraint is unknown" should say "if there is no constraining mechanism, or the constraint is unknown" but there is a mechanism in effect and the constraint is known, it's set to the max.

I will say that Node.js currently gets it wrong. It seems to assume that if `uv_get_constrained_memory()` returns non-zero, that implicitly means it's less than `uv_get_free_memory()` or `uv_get_total_memory()`. Not strictly always true.

Comment (maleadt) [reply src/unix/linux-core.c:804]: Is it OK to mention the return value for no limit/max in the documentation? I've updated https://github.com/libuv/libuv/pull/3753.

**Pull request merged**
